src/init.cpp

```diff
@@ -1686,27 +1686,18 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
                 break;
             }
 
-            bool failed_rewind{false};
-            // Can't hold cs_main while calling RewindBlockIndex, so retrieve the relevant
-            // chainstates beforehand.
-            for (CChainState* chainstate : WITH_LOCK(::cs_main, return chainman.GetAll())) {
-                if (!fReset) {
-                    // Note that RewindBlockIndex MUST run even if we're about to -reindex-chainstate.
-                    // It both disconnects blocks based on the chainstate, and drops block data in
-                    // BlockIndex() based on lack of available witness data.
-                    uiInterface.InitMessage(_("Rewinding blocks...").translated);
-                    if (!chainstate->RewindBlockIndex(chainparams)) {
-                        strLoadError = _(
-                            "Unable to rewind the database to a pre-fork state. "
-                            "You will need to redownload the blockchain");
-                        failed_rewind = true;
-                        break; // out of the per-chainstate loop
-                    }
-                }
-            }
 
-            if (failed_rewind) {
-                break; // out of the chainstate activation do-while
+            if (!fReset) {
+                uiInterface.InitMessage(_("Checking if redownload is needed...").translated);
+                LOCK(cs_main);
+                auto chainstates{chainman.GetAll()};
+                if (std::any_of(chainstates.begin(), chainstates.end(),
+                                [&chainparams](CChainState* cs) EXCLUSIVE_LOCKS_REQUIRED(cs_main) { return cs->NeedsRedownload(chainparams); })) {
+                    strLoadError = _(
+                        "Segwit blocks are insufficiently validated. "
+                        "Please delete blocks dir and chain state dir and restart.");
```

Comment:
Should we be more cautious with directly summoning users to delete directories ?

Maybe we should give a more detailed path (`.bitcoin/blocks`, `.bitcoin/chainstate`) but not sure if it's platform dependent. I just want to minimize the risk of someone deleting by mistake its wallet.

Reply:
Good idea! Done.